Add script to find terms without an example

bact opened the pull request:
# Pull Request

From DPV 2.3, new concepts should have an example (see 13 Aug 2025 meeting).

This `295_find_terms_without_example.py` script will find terms without an example in the `examples` directory.

Update: The script will also find terms used in an example but undefined in vocab files (a potential typo).


Running without option at command line, it will show numbers of terms with an example (or that one of their parents has an example):

```text
Namespace                  Class w/ Examples Prop. w/ Examples
-------------------------- ----------------- -----------------
ai                               4 / 183           1 / 10     
dpv                            442 / 962         113 / 144    
eu-aiact                         7 / 105           2 / 2      
eu-dga                          20 / 62            5 / 5      
eu-ehds                         17 / 61            0 / 0      
eu-gdpr                         87 / 217           6 / 6      
eu-nis2                          0 / 12            0 / 0      
eu-rights                        1 / 137           0 / 0      
justifications                  20 / 66            0 / 0      
legal-eu                         8 / 21            0 / 0      
loc                            146 / 5270          0 / 4      
p7012                           38 / 136           4 / 21     
pd                              70 / 221           0 / 0      
risk                            62 / 491          13 / 43     
sector-education                 7 / 49            0 / 0      
sector-finance                  10 / 43            0 / 0      
sector-health                    3 / 59            0 / 0      
sector-infra                     1 / 47            0 / 0      
sector-law                       2 / 94            0 / 0      
sector-publicservices            3 / 15            0 / 0      
tech                            14 / 127           5 / 52     

Top parents among classes without examples (excluding 'loc:'):
     49  risk:RiskMatrix7x7
     25  risk:RiskMatrix5x5
     25  risk:ServiceRelatedConsequence
     24  justifications:LegalProcessImpaired
     20  tech:Actor
     17  dpv:CryptographicMethods
     16  dpv:SecurityMethod
     16  risk:Discrimination
     16  dpv:PublicBenefit
     16  dpv:DataTransferLegalBasis

Top parents among properties without examples:
     18  tech:hasActor
      6  risk:controls
      5  dpv:hasData
      4  ai:hasAI
      4  skos:altLabel
      3  tech:hasInputData
      3  ai:hasData
      3  risk:resolves
      2  risk:reduces
      2  tech:hasInput  
```

Running with `-v` option, it will print all the terms without an example:

```text
==== Properties without examples ====

ai:hasAISystem                           ⊂ ai:hasAI
ai:hasCapability                         ⊂ ai:hasAI
ai:hasData                               ⊂ dpv:hasData
ai:hasGPAIModel                          ⊂ ai:hasModel
ai:hasModel                              ⊂ ai:hasAI
ai:hasTechnique                          ⊂ ai:hasAI
ai:hasTestingData                        ⊂ ai:hasData, tech:hasInputData
ai:hasTrainingData                       ⊂ ai:hasData, tech:hasInputData
ai:hasValidationData                     ⊂ ai:hasData, tech:hasInputData
dpv:hasConformanceStatus
dpv:hasData
dpv:hasDataSubjectScale                  ⊂ dpv:hasScale
...
```

Not exactly useful yet since it can't distinguished the new terms (from one version to another).
Eventually, once we have a "sinceVersion" information (see #359), we may able to show only new terms without an example.

bact committed:
Create 295_find_terms_without_example.py

Find terms without an example in examples directory

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

coolharsh55 commented:
We can use #12 for the general discussion on what to use for example / use-case and how to script tests around it.

bact commented:
> This is cool, I'll run it on 2.2 later. I think this is taking all the RDF outputs and checking which terms occur (anywhere) in examples?

Yes. Since the TTLs in examples/ directory are not in full form, I just match terms with a regular expression (without actual Turtle/RDF parsing). Will put this in code comment.

I will check the `dex`. I have look at it before but only see a description and a link to an actual TTL example file, so I use TTLs in /examples/ (at root) instead. Will look at the code again.

coolharsh55 commented:
What do you mean by full form? They should be valid as turtle - except the name spaces which are taken from the csvs.

bact committed:
Add comments

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

coolharsh55 commented:
I see. It's possible to make them fully conformant turtle files. My worry was that this might take up too much space in the html, but I can truncate the namespaces there via code. Please open an issue for this and I'll implement it later. Though do we need this for v2.2 or can it be done for v2.3? I prefer later as this might break stuff.

bact committed:
Allow to set vocab_dir and examples_dir from command line

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Print top parents

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Print terms found in examples but not defined

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Print filename of undefined terms in examples

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add odrl prefix to skip list

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Search faster

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

coolharsh55 commented:
Thanks, I'll change what I can find from these. There is code somewhere in the existing setup that distinguishes between "DPV concepts" and others based on namespaces in order to generate RDF or HTML. Later when looking at this, I'll see how to use that here as then we won't have to rely on regex and broken/invalid RDF will also be flagged automatically.

coolharsh55 commented:
Thanks @bact -- added the changes in https://github.com/w3c/dpv/commit/22222befed48eb5cd9f8df60eef795cbc366a27c Can you please mark the unresolved ones, or ideally move them to #372 and I'll take a look.

bact commented:
I will move the unresolved ones to #372 to better track that.

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Exclude -owl.ttl

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Find terms referenced in HTML but undefined

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Fix quote in quote

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add extract term in backtick

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact committed:
Add comment

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact commented:
The numbering of the script (2xx) is a bit confusing though since there is one part (the check of possible undefined terms in HTML) that should be run after the HTML generation script (300). 

Can think more about this for 2.3.

coolharsh55 commented:
@bact How about -- we make all `4xx` numbered scripts be for testing -- so any tests go there, including the current `290` for SHACL and future ones like a fork of OOPS!/FOOPS! specifically for DPV that I'm planning to write. The current logic is that `1xx` is data retrieval from GSheets, `2xx` is RDF output, and `3xx` is HTML output, then `9xx` is for releases. In the future, we will need further numbers for tools/implementations e.g. if we host something interactive or provide a library or something -- these can take up `5xx` -- `8xx`.

bact commented:
What about 

- 1xx data retrieval from source
- 2xx RDF generation (for machine)
- 3xx HTML generation (for human)
- 4xx Test
  - 41x Test retrieved data
  - 42x Test generated RDF
  - 43x Test generated HTML 

But this will make the possible numbers for each category limited to only 9 and difficult to allocate in a forward-compatible way (keep same numbers in the future).

coolharsh55 commented:
Isn't the issue here that the script tests both RDF and HTML -- and then whether it should be `42x` or `43x`? I think the `4xx` should not follow the numbering e.g. what if we have more than 10 RDF test files. There very likely won't be a lot of code here and we can put the workflow in the README or wiki if it isn't simple/intuitive. So the primary focus should be of usability for people working with/on it (e.g. you and me).

bact commented:
You're right. Keep it 4xx simple. Thanks.

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Fix example term regex

Do not match strings (surrounded with quotes) and datetime-lookalike

Co-Authored-By: Harshvardhan Pandit <me@harshp.com>

bact commented:
> Tested this for fixing typos, brilliant stuff @bact - very helpful! Some minor issues which are easily recognised and ignored:
> 
> ```
> 2021-05-28T12:24                         in: E0023.ttl <--- ignore strings i.e. starting with ""
> 2022-09-06T15:36                         in: E0016.ttl <--- ignore strings i.e. starting with ""
> dpv:DataLoss                             in: E0068.ttl, E0069.ttl
> eu-gdpr:A6                               in: E0072.ttl <--- terms can contain dashes e.g. A6-xyz
> eu-gdpr:A7                               in: E0072.ttl <--- terms can contain dashes e.g. A7-xyz
> exA:TVServiceOptimisation                in: E0004.ttl <--- example prefixes can be of the form ex, exA, ... exF
> exB:TVServiceOptimisation                in: E0004.ttl <--- example prefixes can be of the form ex, exA, ... exF
> exB:TVSignalOptimisation                 in: E0004.ttl <--- example prefixes can be of the form ex, exA, ... exF
> legal-eu:law                             in: E0036.ttl, E0055.ttl, E0067.ttl, E0072.ttl, E0076.ttl <--- terms can contain dashes
> legal-ie:DPA                             in: E0036.ttl <--- terms can contain dashes
> legal-ie:law                             in: E0036.ttl <--- terms can contain dashes
> nace:72                                  in: E0008.ttl <--- to add nace prefix in our list
> policy:1                                 in: E0030.ttl <--- syntax error in IRI, fixed
> ```

Fixed the regex as suggested.

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact commented:
Run against 2.3-dev on 2026-02-04 11:15 UTC

```text
Classes (inc. parents) with examples: 1091 / 8487
Properties (inc. parents) with examples: 150 / 287

Namespace                  Class w/ Examples Prop. w/ Examples
-------------------------- ----------------- -----------------
ai                               5 / 185           1 / 10     
dpv                            448 / 962         113 / 144    
eu-aiact                         7 / 105           2 / 2      
eu-dga                          20 / 62            5 / 5      
eu-ehds                         17 / 61            0 / 0      
eu-gdpr                         94 / 217           6 / 6      
eu-nis2                          0 / 12            0 / 0      
eu-rights                        1 / 137           0 / 0      
justifications                  20 / 66            0 / 0      
legal-eu                         8 / 21            0 / 0      
loc                            146 / 5270          0 / 4      
p7012                           38 / 136           4 / 21     
pd                              70 / 221           0 / 0      
risk                            71 / 492          14 / 43     
sector-education                 7 / 49            0 / 0      
sector-finance                  10 / 43            0 / 0      
sector-health                    3 / 59            0 / 0      
sector-infra                     1 / 47            0 / 0      
sector-law                       2 / 94            0 / 0      
sector-publicservices            3 / 15            0 / 0      
tech                            14 / 127           5 / 52  

Top parents among classes without examples
(excluding child with prefixes: loc)
-------------------------------------------------------------
     49  risk:RiskMatrix7x7
     25  risk:ServiceRelatedConsequence
     25  risk:RiskMatrix5x5
     24  justifications:LegalProcessImpaired
     20  tech:Actor
     17  dpv:CryptographicMethods
     16  dpv:PublicBenefit
     16  risk:Discrimination
     16  dpv:SecurityMethod
     15  dpv:DataTransferLegalBasis

Top parents among properties without examples
-------------------------------------------------------------
     18  tech:hasActor
      6  risk:controls
      5  dpv:hasData
      4  ai:hasAI
      4  skos:altLabel
      3  ai:hasData
      3  tech:hasInputData
      3  risk:resolves
      2  tech:hasInput
      2  dpv:hasTechnicalOrganisationalMeasure

Terms used in examples but NOT defined in vocabulary files (1)
----------------------------------------------------------
nace:72                                  in: E0008.ttl

Terms referenced in HTML but NOT defined in vocabulary files (0)
------------------------------------------------------------
Not found.
```

coolharsh55 commented:
@bact 2.3 is now released and I want to add this to 2.4 -- let me know what's the best way to proceed (e.g. do I just merge and use this, should we meet and discuss it). Thanks.

bact committed:
Update DEFAULT_VOCAB_DIR to 2.3

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>

bact commented:
I think the script itself currently workable for our directory structure, so probably ready to merge after renaming the script to make it fit numbering scheme discussed above.

Maybe if you can help review the command line options and their description if they make sense and will not confused/give wrong assumption to people, that will be great.

Anything other than the script naming and the command line options (the two that may affect other scripts/documentation), like additional features or relevant namescape updates can be made in another PR.

Any particular types of check/validation do you feel DPV need the most and it is missing at this moment?

coolharsh55 commented:
Thanks, I will add the options as a separate file alongside existing SHACL shapes. I don't think anything is missing in particular. We could do something similar for use-cases in the future, but let's get examples done first.

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact committed:
Merge branch 'dev' into find-terms-without-examples

bact commented on code/400_find_terms_without_example.py:
```diff
@@ -0,0 +1,492 @@
+    "xsd",
+}
+
+DEFAULT_VOCAB_DIR = "../2.3"
```
Update this version number.

bact committed:
Update DEFAULT_VOCAB_DIR to version 2.4-dev

Signed-off-by: Arthit Suriyawongkul <arthit@gmail.com>
